lang/java/compiler/src/main/java/org/apache/avro/compiler/specific/SpecificCompiler.java

```diff
@@ -270,17 +270,22 @@ File writeToDestination(File src, File destDir) throws IOException {
       if (src != null && f.exists() && f.lastModified() >= src.lastModified())
         return f;                                 // already up to date: ignore
       f.getParentFile().mkdirs();
-      Writer fw;
-      if (outputCharacterEncoding != null) {
-        fw = new OutputStreamWriter(new FileOutputStream(f), outputCharacterEncoding);
-      } else {
-        fw = new FileWriter(f);
-      }
+      Writer fw = null;
+      FileOutputStream fos = null;
       try {
+        if (outputCharacterEncoding != null) {
+          fos = new FileOutputStream(f);
+          fw = new OutputStreamWriter(fos, outputCharacterEncoding);
+        } else {
+          fw = new FileWriter(f);
+        }
         fw.write(FILE_HEADER);
         fw.write(contents);
       } finally {
-        fw.close();
+        if (fw != null)
+          fw.close();
+        if (fos != null)
+          fos.close();
```

Comment:
fos.close() is invoked separately to close the file in case of the fw creation fails (fw starts writing at the constructor). If both fos and fw creation was successful than a problem occurs during writing fw.close() shall close the file. If it fails, fos.close() would probably fail too.